tools/src/describe.rs

```diff
@@ -0,0 +1,97 @@
+use std::ffi::OsStr;
+use std::path::Path;
+
+use languages::FormatAs;
+
+use regex::Regex;
+
+/// Tries to find some text that describes the contents of the file.
+/// Adapted from DXR code at https://github.com/mozilla/dxr/blob/master/dxr/plugins/descriptor/__init__.py
+pub fn describe_file(contents: &str, path: &Path, format: &FormatAs) -> Option<String> {
+    // Only look in the first 5k chars of the file, since running regex matches on giant files
+    // can be expensive and most likely the thing we're looking for will be early in the file.
+    let substr_end = contents.char_indices().nth(5000).map(|ix| ix.0).unwrap_or(contents.len());
+    let substr = &contents[0..substr_end];
+
+    // DXR also does a search for "filename: <description>" which I've never seen in any file so
+    // I'm omitting that here. We can add it if needed.
+    match format {
+        FormatAs::FormatTagLike(_) => describe_html(substr),
+        FormatAs::FormatCLike(ref spec) if spec.c_style_comments => describe_from_c_comment(substr),
+        FormatAs::FormatCLike(ref spec) if spec.triple_quote_literals => describe_py(substr),
+        FormatAs::FormatCLike(_) | FormatAs::Binary => None,
+        FormatAs::Plain => {
+            if path.file_stem().and_then(OsStr::to_str).map(|s| s.to_uppercase() == "README").unwrap_or(false) {
```

Comment:
nit: This may be nicer like:

```rust
let stem = path.file_stem()?.to_str()?;
if stem.eq_ignore_ascii_case("README") {
    describe_readme(substr)
} else {
    None
}
```

Or something of that sort.

Reply:
That is much nicer, particularly with `eq_ignore_ascii_case` which I forgot about. Thanks!